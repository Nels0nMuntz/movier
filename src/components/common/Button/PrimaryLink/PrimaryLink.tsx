import { styled } from "@mui/material/styles";
import React, { FC, PropsWithChildren } from "react";
import { Typography } from "../../Typography/Typography";
import { Link } from "react-router-dom";


export interface Props extends PropsWithChildren {
  href: string;
  icon?: React.ReactNode;
  children?: string;
}

const StyledLink = styled(Link)(({ theme }) => ({
  position: "relative",
  display: "flex",
  alignItems: "center",
  gap: theme.spacing(1),
  paddingLeft: theme.spacing(3),
  paddingRight: theme.spacing(3),
  paddingTop: theme.spacing(1),
  paddingBottom: theme.spacing(1),
  background: theme.palette.secondary.main,
  "& *": {
    position: "relative",
    zIndex: 1,
  },
  "&:before": {
    position: "absolute",
    content: "\"\"",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    background: "#333",
    transition: "transform 0.3s ease-out",
    transform: "scaleX(0)",
    transformOrigin: "0 50%",
  },
  "&:hover": {
    background: theme.palette.common.black,
    "&:before": {
      transform: "scaleX(1)",
    }
  }
}));

export const PrimaryLink: FC<Props> = ({ href, icon, children }) => {
  return (
    <StyledLink to={href}>
      {icon}
      {children && (
        <Typography element="span" type="body_1" data-testid="primary-link-children">
          {children}
        </Typography>
      )}
    </StyledLink>
  )
};