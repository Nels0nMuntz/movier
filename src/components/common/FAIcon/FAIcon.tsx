import { FC } from "react";
import { IconProp } from "@fortawesome/fontawesome-svg-core"
import { FontAwesomeIcon, FontAwesomeIconProps,  } from "@fortawesome/react-fontawesome"


interface Props {
  icon: IconProp;
  color?: string;
}

export const FAIcon: FC<Props & FontAwesomeIconProps> = ({ icon, color, ...rest }) => {
  return (
    <FontAwesomeIcon 
      icon={icon} 
      color={color || "white"}
      {...rest}
    />
  )
};