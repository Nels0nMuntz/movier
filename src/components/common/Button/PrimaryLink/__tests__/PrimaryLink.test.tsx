import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { FAIcon } from "components/common/FAIcon/FAIcon";
import { faAd } from "@fortawesome/free-solid-svg-icons";

import { PrimaryLink, Props } from "../PrimaryLink";


describe("PrimaryLink", () => {
  let props: Partial<Props> = {};

  beforeEach(() => {
    props = {
      href: "/",
      children: "Link",
    }
  });

  const renderElement = () => {
    return render(
      <BrowserRouter>
        <PrimaryLink {...props as Props} />
      </BrowserRouter>
    );
  };

  it("renders correctly", () => {
    expect(renderElement().getByRole("link")).toMatchSnapshot();
  });

  it("renders `href` attribute correctly", () => {
    expect(renderElement().getByRole("link")).toHaveAttribute("href", props.href)
  });

  describe("when `children` prop is passed", () => {
    beforeEach(() => {
      props.children = "Link with children";
    });
    it("renders children", () => {
      expect(renderElement().getByText(props.children as string)).toBeInTheDocument()
    })
  });
  describe("when `children` prop is not passed", () => {
    beforeEach(() => {
      props.icon = undefined;
      props.children = undefined;
    });
    it("does not render children", () => {
      expect(renderElement().queryByTestId("primary-link-children")).toBeNull()
    })
  });
  describe("when `icon` prop is passed", () => {
    beforeEach(() => {
      props.icon = (
        <span data-testid="primary-link-icon">
          <FAIcon icon={faAd}/>
        </span>
      )
    });
    it("renders icon", () => {
      expect(renderElement().getByTestId("primary-link-icon")).toBeInTheDocument()
    })
  });
})