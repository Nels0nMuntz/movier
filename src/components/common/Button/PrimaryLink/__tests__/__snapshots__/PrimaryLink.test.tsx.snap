// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryLink renders correctly 1`] = `
<a
  class="css-eos4vf"
  href="/"
>
  <span
    class="css-1n2b22m"
    data-testid="primary-link-children"
    data-typography-type="body_1"
  >
    Link
  </span>
</a>
`;
